### PR TITLE
Factory methods for UUID-based `Message` identifiers

### PR DESCRIPTION
**Pull request by dmitrykuzmin**
This PR resolves issue https://github.com/SpineEventEngine/core-java/issues/778.

The `MarkerInterfaceGenerator` will search for messages like:

```
message ProjectId {
    string uuid = 1;
}
```
i.e. with a single `string` value named "uuid" and mark them as `UniqueId`.

The `UniqueId` interface provides convenient `default` methods for messages creation: `generate` and `of(String)` to generate an ID from random string or given value respectively.

Also the `MarkerInterfaceGenerator` was extracted into a separate package `marker` under the `io.spine.tools.protoc`.

### CODE REVIEW

**Comment by dmitrykuzmin** [comment]
@armiol PTAL.

File: tools/protoc-plugin/src/main/java/io/spine/tools/protoc/MarkerInterfaceParameter.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * The generic parameter of the {@link MarkerInterface}.
+ *
+ * <p>Parameter value is presented as a {@code String} for usage in the generated code.
+ */
+interface MarkerInterfaceParameter {
+
+    /**
+     * Obtains a parameter value based on who is the marker interface implementor.
+     *
+     * @param implementor
```
Comment (armiol) [inline tools/protoc-plugin/src/main/java/io/spine/tools/protoc/MarkerInterfaceParameter.java]: Let's call it `descendant` or `target` (i.e. a target to mark).

File: tools/protoc-plugin/src/main/java/io/spine/tools/protoc/MessageItself.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.google.protobuf.DescriptorProtos.DescriptorProto;
+
+/**
+ * The marker interface parameter whose value is the implementing {@code Message} itself.
+ *
+ * <p>For example, for the {@code ProjectId} class, the value of the parameter will be:
+ * {@code ProjectId}.
+ */
+final class MessageItself implements MarkerInterfaceParameter {
```
Comment (armiol) [inline tools/protoc-plugin/src/main/java/io/spine/tools/protoc/MessageItself.java]: Let's discuss. 

To me a class such as `io.spine.tools.protoc.MessageItself` is questionable.

Let's consider splitting marker interface-specific routines into their own package.

File: tools/protoc-plugin/src/test/proto/spine/tools/protoc/incorrect_ids.proto [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import "spine/options.proto";
+
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.tools.protoc";
+option java_outer_classname = "IncorrectIds";
+option java_multiple_files = true;
+
+import "google/protobuf/timestamp.proto";
+
+message NotId {
```
Comment (armiol) [inline tools/protoc-plugin/src/test/proto/spine/tools/protoc/incorrect_ids.proto]: Same as above. It _is_ an ID. Or at least there can be an identifier like this.

File: tools/protoc-plugin/src/test/proto/spine/tools/protoc/incorrect_ids.proto [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+//
+syntax = "proto3";
+
+package spine.tools.protoc;
+
+import "spine/options.proto";
+
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.tools.protoc";
+option java_outer_classname = "IncorrectIds";
```
Comment (armiol) [inline tools/protoc-plugin/src/test/proto/spine/tools/protoc/incorrect_ids.proto]: Can we have the naming of the outer class and the file somehow related to `UniqueId` marker interface?

File: base/src/main/java/io/spine/base/UniqueId.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+     * Generates a new identifier instance using a random {@code String}.
+     */
+    default I generate() {
+        Class<I> thisClass = (Class<I>) this.getClass();
+        UuidMessage<I> uuidMessage = UuidMessage.of(thisClass);
+        return uuidMessage.generate();
+    }
+
+    /**
+     * Generates a new identifier instance from the passed value.
```
Comment (alexander-yevsyukov) [inline base/src/main/java/io/spine/base/UniqueId.java]: It does not generate. It creates an instance by value.

File: base/src/main/java/io/spine/base/UniqueId.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ *     }
+ * </pre>
+ *
+ * @param <I>
+ *         the type of the message
+ */
+@SuppressWarnings({"unchecked" /* Class cast correctness guaranteed by the interface contract. */,
+        "InterfaceNeverImplemented" /* Used by the Protobuf Compiler plugin. */})
+@Immutable
+public interface UniqueId<I extends Message> extends SerializableMessage {
```
Comment (alexander-yevsyukov) [inline base/src/main/java/io/spine/base/UniqueId.java]: Why do we want `Unique` in the name? What's so unique about this name?

File: base/src/main/java/io/spine/base/UniqueId.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        return uuidMessage.generate();
+    }
+
+    /**
+     * Generates a new identifier instance from the passed value.
+     */
+    default I of(String value) {
+        Class<I> thisClass = (Class<I>) this.getClass();
+        UuidMessage<I> uuidMessage = UuidMessage.of(thisClass);
+        return uuidMessage.create(value);
```
Comment (alexander-yevsyukov) [inline base/src/main/java/io/spine/base/UniqueId.java]: What happens if the passed value is not a UUID-based string?

**Comment by alexander-yevsyukov** [comment]
How about renaming `UuidMessage` into `UuidFactory` and having `UuidValue` as an interface (which is now named `UniqueId`)?

**Comment by dmitrykuzmin** [comment]
@armiol PTAL.

**Comment by dmitrykuzmin** [comment]
@armiol PTAL.

**Pull request merged**